bob-apps/bin/bobc.rs

```diff
@@ -1,71 +1,365 @@
+use clap::{App, AppSettings, Arg, ArgMatches, SubCommand};
 use http::Uri;
+use lazy_static::lazy_static;
 use log::LevelFilter;
+use regex::Regex;
 use std::fmt::Debug;
+use std::ops::RangeInclusive;
+use std::path::{Path, PathBuf};
 use std::str::FromStr;
 use std::time::{SystemTime, UNIX_EPOCH};
+use tokio::fs::{read_dir, File};
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
 use tonic::Request;
 
+struct NetConfig {
+    port: u16,
+    target: String,
+}
+impl NetConfig {
+    fn get_uri(&self) -> http::Uri {
+        http::Uri::builder()
+            .scheme("http")
+            .authority(format!("{}:{}", self.target, self.port).as_str())
+            .path_and_query("/")
+            .build()
+            .unwrap()
+    }
+
+    fn from_args(args: &AppArgs) -> Self {
+        Self {
+            port: args.port,
+            target: args.host.clone(),
+        }
+    }
+}
+
+const KEY_ARG: &str = "key";
+const KEY_SIZE_ARG: &str = "key-size";
+const HOST_ARG: &str = "host";
+const PORT_ARG: &str = "port";
+const FILE_ARG: &str = "file";
+
+const PUT_SC: &str = "put";
+const GET_SC: &str = "get";
+const EXISTS_SC: &str = "exists";
+
+#[derive(Debug)]
+enum ParseError {
+    FilePattern,
+    KeyPattern,
+}
+
+#[derive(Debug)]
+struct AppArgs<'a> {
+    subcommand: &'a str,
+    file_pattern: Option<FilePattern<'a>>,
+    key_pattern: Option<KeyPattern>,
+    keysize: usize,
+    host: String,
+    port: u16,
+}
+
+impl<'a> AppArgs<'a> {
+    fn from_matches(matches: &'a ArgMatches<'a>) -> Self {
+        let (subcommand, sub_matches) = matches.subcommand();
+        let sub_matches = sub_matches.unwrap();
+        let file_pattern = match sub_matches.value_of(FILE_ARG) {
+            None => None,
+            Some(p) => match Self::parse_file_pattern(p) {
+                Ok(val) => Some(val),
+                Err(e) => panic!("{:?}", e), // should send an error to the caller
+            },
+        };
+        let key_pattern = match sub_matches.value_of(KEY_ARG) {
+            None => None,
+            Some(p) => match Self::parse_key_pattern(p) {
+                Ok(val) => Some(val),
+                Err(e) => panic!("{:?}", e), // again
+            },
+        };
+
+        Self {
+            subcommand,
+            file_pattern,
+            key_pattern,
+            keysize: sub_matches.value_or_default(KEY_SIZE_ARG),
+            host: sub_matches.value_or_default(HOST_ARG),
+            port: sub_matches.value_or_default(PORT_ARG),
+        }
+    }
+    fn parse_file_pattern(file_pattern: &str) -> Result<FilePattern<'_>, ParseError> {
+        let path = PathBuf::from(file_pattern);
+        let name = path.file_name().unwrap().to_str().unwrap();
+
+        let re = Regex::new("\\{key\\}").unwrap();
+        let two_or_more = Regex::new(".*\\{key\\}.*\\{key\\}.*").unwrap();
+
+        if re.is_match(name) {
+            if two_or_more.is_match(name) {
```

Comment:
You can use the same `re` regex for this check: `re.find_iter(..).count()` will give the number of matches

Reply:
Fix: [f7d82a2](https://github.com/qoollo/bob/pull/562/commits/f7d82a260c1d009f34de17d98ce505a63b5810b0)